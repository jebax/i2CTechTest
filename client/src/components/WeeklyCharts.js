import React, { Component } from 'react'
import axios from 'axios'
import '../../node_modules/react-vis/dist/style.css'
import ExposedControlChart from './ExposedControlChart'

export default class WeeklyCharts extends Component {
  state = {
    aisleExposed: [],
    aisleControl: [],
    brandExposed: [],
    brandControl: [],
    offerExposed: [],
    offerControl: []
  }

  componentDidMount() {
    axios.get(process.env.REACT_APP_WEEKLY_URL).then(response => {
      this.formatData(response.data)
    })
  }

  formatData = data => {
    let aisleExposed = []
    let aisleControl = []

    let brandExposed = []
    let brandControl = []

    let offerExposed = []
    let offerControl = []

    data.forEach(entry => {
      if (entry.PRODUCT === 'Aisle') {
        this.appendFormattedData(aisleExposed, aisleControl, entry)
      } else if (entry.PRODUCT === 'Brand') {
        this.appendFormattedData(brandExposed, brandControl, entry)
      } else {
        this.appendFormattedData(offerExposed, offerControl, entry)
      }
    })

    this.setState(prevState => ({
      aisleExposed: aisleExposed,
      aisleControl: aisleControl,
      brandExposed: brandExposed,
      brandControl: brandControl,
      offerExposed: offerExposed,
      offerControl: offerControl
    }))
  }

  appendFormattedData = (exposedCollection, controlCollection, entry) => {
    exposedCollection.push(
      {
        x: new Date(entry.WEEK_COMMENCING),
        y: entry.EXPOSED
      }
    )
    controlCollection.push(
      {
        x: new Date(entry.WEEK_COMMENCING),
        y: entry.CONTROL
      }
    )
  }

  render() {
    return(
      <div>
        <ExposedControlChart
          exposed={this.state.aisleExposed}
          control={this.state.aisleControl}
        />
        <ExposedControlChart
          exposed={this.state.brandExposed}
          control={this.state.brandControl}
        />
        <ExposedControlChart
          exposed={this.state.offerExposed}
          control={this.state.offerControl}
        />
      </div>
    )
  }
}
