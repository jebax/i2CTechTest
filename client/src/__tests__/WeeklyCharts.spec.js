import React from 'react'
import { shallow, mount } from 'enzyme'
import axios from 'axios'
import WeeklyCharts from '../components/WeeklyCharts'

jest.mock('axios')

describe('Weekly charts wrapper', () => {
  var weeklyCharts
  var weeklyData

  beforeAll(() => {
    weeklyData = [
      {
        PRODUCT: 'Aisle',
        WEEK_COMMENCING: '2018-01-01',
        EXPOSED: '222',
        CONTROL: '111'
      },
      {
        PRODUCT: 'Brand',
        WEEK_COMMENCING: '2018-01-01',
        EXPOSED: '444',
        CONTROL: '333'
      },
      {
        PRODUCT: 'Offer',
        WEEK_COMMENCING: '2018-01-01',
        EXPOSED: '666',
        CONTROL: '555'
      }
    ]

    axios.get.mockResolvedValue({ data: weeklyData })

    weeklyCharts = shallow(<WeeklyCharts />)
  })

  it('renders correctly', () => {
    expect(weeklyCharts).toMatchSnapshot()
  })

  it('makes the correct API request', () => {
    expect(axios.get).toHaveBeenCalledTimes(1)
    expect(axios.get).toHaveBeenCalledWith(process.env.REACT_APP_WEEKLY_URL)
  })

  it('includes three ExposedControlCharts', () => {
    expect(weeklyCharts.find('ExposedControlChart').length).toEqual(3)
  })

  it('passes the aisle data to the first ExposedControlChart', () => {
    const aisleChart = weeklyCharts.find('ExposedControlChart').first()

    const exposedData = [{ x: new Date('2018-01-01'), y: '222' }]
    const controlData = [{ x: new Date('2018-01-01'), y: '111' }]

    expect(aisleChart.props().exposed).toEqual(exposedData)
    expect(aisleChart.props().control).toEqual(controlData)
  })

  it('passes the brand data to the second ExposedControlChart', () => {
    const brandChart = weeklyCharts.find('ExposedControlChart').at(1)

    const exposedData = [{ x: new Date('2018-01-01'), y: '444' }]
    const controlData = [{ x: new Date('2018-01-01'), y: '333' }]

    expect(brandChart.props().exposed).toEqual(exposedData)
    expect(brandChart.props().control).toEqual(controlData)
  })

  it('passes the offer data to the third ExposedControlChart', () => {
    const offerChart = weeklyCharts.find('ExposedControlChart').at(2)

    const exposedData = [{ x: new Date('2018-01-01'), y: '666' }]
    const controlData = [{ x: new Date('2018-01-01'), y: '555' }]

    expect(offerChart.props().exposed).toEqual(exposedData)
    expect(offerChart.props().control).toEqual(controlData)
  })
})
