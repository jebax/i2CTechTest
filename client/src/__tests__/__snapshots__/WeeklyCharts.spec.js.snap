// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weekly charts wrapper renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeeklyCharts />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ExposedControlChart
          control={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "111",
              },
            ]
          }
          exposed={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "222",
              },
            ]
          }
        />,
        <ExposedControlChart
          control={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "333",
              },
            ]
          }
          exposed={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "444",
              },
            ]
          }
        />,
        <ExposedControlChart
          control={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "555",
              },
            ]
          }
          exposed={
            Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "666",
              },
            ]
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "control": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "111",
            },
          ],
          "exposed": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "222",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "control": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "333",
            },
          ],
          "exposed": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "444",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "control": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "555",
            },
          ],
          "exposed": Array [
            Object {
              "x": 2018-01-01T00:00:00.000Z,
              "y": "666",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ExposedControlChart
            control={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "111",
                },
              ]
            }
            exposed={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "222",
                },
              ]
            }
          />,
          <ExposedControlChart
            control={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "333",
                },
              ]
            }
            exposed={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "444",
                },
              ]
            }
          />,
          <ExposedControlChart
            control={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "555",
                },
              ]
            }
            exposed={
              Array [
                Object {
                  "x": 2018-01-01T00:00:00.000Z,
                  "y": "666",
                },
              ]
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "control": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "111",
              },
            ],
            "exposed": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "222",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "control": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "333",
              },
            ],
            "exposed": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "444",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "control": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "555",
              },
            ],
            "exposed": Array [
              Object {
                "x": 2018-01-01T00:00:00.000Z,
                "y": "666",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
